### PR TITLE
NF: add dipy_math workflow

### PR DESCRIPTION
**Pull request by skoudoro**
this PR adds a new CLI named `dipy_math` which allows to do basic operations on volume (nifti images).

suppose you have 1, 2 or 3 nifti volumes in a folder, you could do:
- `dipy_math "vol1 + vol2" nifti_a.nii.gz nifti_b.nii.gz`
- `dipy_math "vol1 - vol2 * vol3" nifti_a.nii.gz nifti_b.nii.gz nifti_c.nii.gz`
- `dipy_math "vol1 / vol2" nifti_a.nii.gz nifti_b.nii.gz -o my_result.nii.gz` 
- `dipy_math  "sin(vol1) + arcsinh(vol2/vol3)" nifti_a.nii.gz nifti_b.nii.gz nifti_c.nii.gz -o my_result.nii.gz`
- etc.

### CODE REVIEW

**Review by maharshi-gor — COMMENTED** [review]
Hey @skoudoro , cool feature!

I have tested this on my local, LGTM! 

Let's add few more corner cases in test to match our criteria for test in the PR.

Thanks!

**Comment by skoudoro** [comment]
> Let's add few more corner cases in test to match our criteria for test in the PR.

Can you be more explicit @maharshi-gor? example of a corner case ?

**Comment by arokem** [comment]
Interesting! I would like to see a bit more error handling. Like when inputs are not matched in terms of their shape, I think it would be good to raise an informative error. I think also the affine needs to be identical, right? Otherwise how would you decide which affine to use?

**Comment by skoudoro** [comment]
> I would like to see a bit more error handling. Like when inputs are not matched in terms of their shape, I think it would be good to raise an informative error. I think also the affine needs to be identical, right? Otherwise how would you decide which affine to use?

ok, I see. I will add that.

Thank you both for the feedback.

**Comment by maharshi-gor** [comment]
> > Let's add few more corner cases in test to match our criteria for test in the PR.
> 
> Can you be more explicit @maharshi-gor? example of a corner case ?

Please check the codecov report it is missing some of the branch coverage, like a few if conditions. For example `dtype` needs to be passed to cover the `if dtype:` line in the coverage.

**Comment by skoudoro** [comment]
> Please check the codecov report it is missing some of the branch coverage, like a few if conditions. For example dtype needs to be passed to cover the if dtype: line in the coverage.

I see, thanks for the feedback @maharshi-gor

**Comment by skoudoro** [comment]
Any additional comment @arokem or @maharshi-gor ?

I think this PR is ready to go and address your last review

File: dipy/workflows/io.py [hunk 1]
```diff
@@ -775,3 +778,125 @@ def run(
...
+            info_msg += f"{fname}:\n- vol index: {i}\n- shape: {data.shape}"
+            info_msg += f"\n- affine:\n{affine}\n"
+            if ref_affine is None:
+                ref_affine = affine
+                ref_shape = data.shape
+                continue
+
+            have_errors = (
+                have_errors
+                or not np.array_equal(ref_affine, affine)
```
Comment (arokem) [inline dipy/workflows/io.py:869]: Do you want to allow for optional tolerance input here in case of affines that differ on a very small number?

File: dipy/workflows/tests/test_io.py [hunk 2]
```diff
@@ -342,3 +349,73 @@ def test_pam_to_niftis_flow():
...
+def test_math():
+    with TemporaryDirectory() as out_dir:
+        data_path, _, _ = get_fnames(name="small_101D")
+        data_path_a = pjoin(out_dir, "data_a.nii.gz")
+        data_path_b = pjoin(out_dir, "data_b.nii.gz")
+        shutil.copy(data_path, data_path_a)
+        shutil.copy(data_path, data_path_b)
+
+        data, _ = load_nifti(data_path)
+        operations = ["vol1*3", "vol1+vol2+vol3", "5*vol1-vol2-vol3", "vol3*2 + vol2"]
```
Comment (arokem) [inline dipy/workflows/tests/test_io.py:363]: Will this elegantly handle extraneous white space like "vol1 + vol2 "?

**Review by arokem — COMMENTED** [review]
A couple more thoughts about this. I think we want this to be pretty bullet proof because it's intended for users who don't program so everything needs to be really clear

File: dipy/workflows/tests/test_io.py (hunk 2 above)
Comment (skoudoro) [reply dipy/workflows/tests/test_io.py:363]: yes, it handles it without problem. I did not put space because 88 line limit 😅

File: dipy/workflows/io.py (hunk 1 above)
Comment (skoudoro) [reply dipy/workflows/io.py:869]: sounds like a good idea, let me update that

File: dipy/workflows/tests/test_io.py (hunk 2 above)
Comment (skoudoro) [reply dipy/workflows/tests/test_io.py:363]: I just realized that the last one on the list has a space between operator

**Comment by skoudoro** [comment]
ok, all has been addressed @arokem and @maharshi-gor. Let me know if I missing something

**Comment by Garyfallidis** [comment]
@arokem is there anything else?
 @deka27 can you check this PR too?

**Review by maharshi-gor — APPROVED** [review]

**Comment by deka27** [comment]
I dont see any issues, nice addition to CLI.
